File: android/vendored/unversioned/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/sensor/ReanimatedSensorContainer.java
package com.swmansion.reanimated.sensor;

import android.util.Log;
import com.facebook.react.bridge.ReactApplicationContext;
import com.swmansion.reanimated.NativeProxy;
import java.lang.ref.WeakReference;
import java.util.HashMap;

public class ReanimatedSensorContainer {

  private int nextSensorId = 0;
  private final WeakReference<ReactApplicationContext> reactContext;
  private final HashMap<Integer, ReanimatedSensor> sensors = new HashMap<>();

  public ReanimatedSensorContainer(WeakReference<ReactApplicationContext> reactContext) {
    this.reactContext = reactContext;
  }

  public int registerSensor(
      ReanimatedSensorType sensorType, int interval, NativeProxy.SensorSetter setter) {
    ReanimatedSensor sensor = new ReanimatedSensor(reactContext, sensorType, interval, setter);
    int sensorId = -1;
    if (sensor.initialize()) {
      sensorId = nextSensorId;
      nextSensorId++;
      sensors.put(sensorId, sensor);
    }
    return sensorId;
  }

  public void unregisterSensor(int sensorId) {
    ReanimatedSensor sensor = sensors.get(sensorId);
    if (sensor != null) {
      sensor.cancel();
      sensors.remove(sensorId);
    } else {
      Log.e("Reanimated", "Tried to unregister nonexistent sensor");
    }
  }
}

File: ios/vendored/unversioned/react-native-reanimated/apple/sensor/ReanimatedSensorContainer.m
#import <Foundation/Foundation.h>
#import <RNReanimated/ReanimatedSensor.h>
#import <RNReanimated/ReanimatedSensorContainer.h>

static NSNumber *_nextSensorId = nil;

@implementation ReanimatedSensorContainer

- (instancetype)init
{
  self = [super init];
  _sensors = [[NSMutableDictionary alloc] init];
  _nextSensorId = @0;
  return self;
}

- (int)registerSensor:(ReanimatedSensorType)sensorType
             interval:(int)interval
    iosReferenceFrame:(int)iosReferenceFrame
               setter:(void (^)(double[], int))setter
{
  ReanimatedSensor *sensor = [[ReanimatedSensor alloc] init:sensorType
                                                   interval:interval
                                          iosReferenceFrame:iosReferenceFrame
                                                     setter:setter];
  if ([sensor initialize]) {
    NSNumber *sensorId = [_nextSensorId copy];
    _nextSensorId = [NSNumber numberWithInt:[_nextSensorId intValue] + 1];
    [_sensors setObject:sensor forKey:sensorId];
    return [sensorId intValue];
  } else {
    return -1;
  }
}

- (void)unregisterSensor:(int)sensorId
{
  NSNumber *_sensorId = [NSNumber numberWithInt:sensorId];
  if (_sensors[_sensorId] == nil) {
    return;
  }
  [_sensors[_sensorId] cancel];
  [_sensors removeObjectForKey:_sensorId];
}

@end
